File: java/algorithm/sort/MergeSort.java
package algorithm.sort;

/**
 * @author roseduan[roseduan520@gmail.com]
 * @time 2019/12/7 21:30
 * @description 归并排序
 **/
public class MergeSort {

    public static void sort(int[] data){
        mergeHelper(data, 0, data.length - 1);
    }

    private static void mergeHelper(int[] data, int p, int r){
        if (p >= r){
            return;
        }
        int q = r + ((p - r) >> 1);

        mergeHelper(data, p, q);
        mergeHelper(data, q + 1, r);
        merge(data, p, q, r);
    }

    private static void merge(int[] data, int p, int q, int r){
        int[] temp = new int[r - p + 1];
        int i = p, j = q + 1, k = 0;

        while (i <= q && j <= r){
            if (data[i] < data[j]){
                temp[k++] = data[i++];
            }
            else {
                temp[k++] = data[j++];
            }
        }

        int start = i;
        int end = q;
        if (j <= r){
            start = j;
            end = r;
        }
        System.arraycopy(data, start, temp, k, end - start + 1);
        System.arraycopy(temp, 0, data, p, temp.length);
    }
}

File: cpp/sort/merge_sort.cpp
//
// Created by roseduan on 2021/9/21.
// 归并排序

#include "vector"
#include "iostream"
using namespace std;

void mergeData(vector<int>& data, int p, int q, int r);
void sortHelper(vector<int>& data, int p, int r);

// 归并排序
void MergeSort(vector<int>& data) {
    sortHelper(data, 0, (int)data.size() - 1);
}

void sortHelper(vector<int>& data, int p, int r) {
    if (p >= r) {
        return;
    }
    int q = p + (r - p) / 2;
    sortHelper(data, p, q);
    sortHelper(data, q + 1, r);
    mergeData(data, p, q, r);
}

void mergeData(vector<int>& data, int p, int q, int r) {
    int i = p, j = q + 1;
    vector<int> temp;
    while (i <= q && j <= r) {
        if (data[i] < data[j]) {
            temp.push_back(data[i++]);
        } else {
            temp.push_back(data[j++]);
        }
    }

    int start = i, end = q;
    if (j <= r) {
        start = j;
        end = r;
    }
    for (int k = start; k <= end; ++k) {
        temp.push_back(data[k]);
    }

    // 拷贝回原数组
    for (int k = p; k <= r; ++k) {
        data[k] = temp[k-p];
    }
}

int main() {
    vector<int> data {1, 0, 5, 9, 6, 5, 1, 0, 4, 7, 3};
    MergeSort(data);

    for (int i : data) {
        cout << i << endl;
    }
}
